### PR TITLE
fix(Script/Quest) Quest 2118 "Plagued Lands"

### LINKED ISSUE #9282: Captured Rabid Thistle Bear of "Plagued Lands" quest is missing disappear flags
### What client do you play on?

enUS

### Faction

- [X] Alliance
- [ ] Horde

### Content Phase:

- [ ] Generic
- [X] 1-19
- [ ] 20-29
- [ ] 30-39
- [ ] 40-49
- [ ] 50-59

### Current Behaviour

Original report: https://github.com/chromiecraft/chromiecraft/issues/1779

I saw a “captured rabid thistle bear” wandering by itself (walking back and forth in a small area) close to the road to the south of Ameth’aran in Darkshore around server time 23:15, 2021.09.17. Checked the same location after the server's daily restart at 05:30, 2021.09.18. The captured rabid thistle bear was replaced by a regular rabid thistle bear. Saw two of them again around server time 03:20, 2021.09.22, to the north of Asterion (the NPC for Bashal'Aran quest series) and the east of Buzzbox 411.
UPDATE: I guess I figured out at least one way to reproduce the issue. See “Extra Notes”.

Tester:

The issue is cofirmed. [Captured Rabid Thistle Bear](https://wowgaming.altervista.org/aowow/?npc=11836) of [Plagued Lands](https://wowgaming.altervista.org/aowow/?quest=2118) quest doesn't dissapear if wasn't taken to quest giver. Rabid Thistle Bears also not respawning while Captured Rabid Thistle Bear is exist. 

https://user-images.githubusercontent.com/85256266/142878586-208ada7f-5c04-4f31-90b9-bcf5a565bd28.mp4

One of tamed by me bear running to spawn location after lost me.

https://user-images.githubusercontent.com/85256266/142881868-3fa6df22-93c9-46a8-bd50-08b29d4bb1e0.mp4

### Expected Blizzlike Behaviour

Running server without crashes/restarts could leave zone without bears. 

### Source

_No response_

### Steps to reproduce the problem

.additem 7586
.go c id 2164
capture a bear
do relog

### Extra Notes

I did some test today:

Step 0. Set hearthstone to Auberdine; wait until it cools down;
Step 1. Accept the quest “plagued lands” from Tharnariun Treetender in Auberdine; find a rabid thistle bear and use the trap provided by the quest NPC to convert the rabid thistle bear to a captured rabid thistle bear;
Step 2. Hearthstone to Auberdine then run back to the spot where the bear is captured; as I approach the spot, the captured bear runs towards me and follows me from there;
Step 3. Abandon the quest; repeat Step 1; instead of hearthstone, just log out and back; find a captured rabid thistle bear wandering by itself;
Step 4.1. Abandon the quest; repeat Step 1; run back to Auberdine or anywhere far away from the spot where the bear is captured (the captured bear should follow);
Step 4.2. Log out and right back in; the bear does not follow any more but instead is walking away; it maybe become invisible (more likely to be underground) or walk in some funny manner;
Step 4.3. Run back to the spot where the bear is captured and wait; eventually the bear will walk back and wander by itself.

So, it seems like I’ve found a method to reliably create “captured rabid thistle bears wandering by itself”. I don’t know whether any other mechanism can also do this. (It would be interesting to test with two chars online at the same time. One of them capture the bear and HS; the other one stands by to see whether the bear will  run/walk towards Auberdine after the first player HS. One player can do this too, if you are willing to wait in Auberdine.)

BTW, as long as a player captured a bear using the trap, it can return to the NPC to turn in the quest. I guess that is what happened: players captured the bears and HS to turn in the quest, and the bears just stay where they are, until a server restart resets them.

### AC rev. hash/commit

https://github.com/chromiecraft/azerothcore-wotlk/commit/7e2e6f8ee85f82f57a3c6e0f6f645b4c42292e7a

### Operating system

Windows 7

### Modules

- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-chromie-xp](https://github.com/azerothcore/mod-chromie-xp)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-ip-tracker](https://github.com/azerothcore/mod-ip-tracker)
- [mod-low-level-arena](https://github.com/azerothcore/mod-low-level-arena)
- [mod-multi-client-check](https://github.com/azerothcore/mod-multi-client-check)
- [mod-pvp-titles](https://github.com/azerothcore/mod-pvp-titles)
- [mod-pvpstats-announcer](https://github.com/azerothcore/mod-pvpstats-announcer)
- [mod-queue-list-cache](https://github.com/azerothcore/mod-queue-list-cache)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-carbon-copy](https://github.com/55Honey/Acore_CarbonCopy)
- [lua-custom-worldboss](https://github.com/55Honey/Acore_CustomWorldboss)
- [lua-level-up-reward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-recruit-a-friend](https://github.com/55Honey/Acore_RecruitAFriend)
- [lua-send-and-bind](https://github.com/55Honey/Acore_SendAndBind)
- [lua-temp-announcements](https://github.com/55Honey/Acore_TempAnnouncements)
- [lua-zonecheck](https://github.com/55Honey/acore_Zonecheck)

### Customizations

None

### Server

ChromieCraft

### PR DESCRIPTION
**Pull request by malcrom**
## Changes Proposed:
-  Replaced SAI with C++ scripting. Quest now works properly

## Issues Addressed:
- Closes #9282

## SOURCE:
Sniff

## Tests Performed:
- Builds
- Tested quest

## How to Test the Changes:
1. take quest from Tharnariun in Aberdine, Darkshore

### CODE REVIEW

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 1]
```diff
@@ -422,10 +422,223 @@ class npc_prospector_remtravel : public CreatureScript
...
+                    {
+                        me->DespawnOrUnsummon();
+                    }
+                    _events.ScheduleEvent(EVENT_CHECK_FOLLOWING, 1000);
+                    break;
+                }
+            }
+
+            if (!UpdateVictim())
+                return;
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:505]: brackets

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 2]
```diff
@@ -422,10 +422,223 @@ class npc_prospector_remtravel : public CreatureScript
...
+                    }
+                    _events.ScheduleEvent(EVENT_POST_QUEST_THREE, 1000);
+                    break;
+                case EVENT_POST_QUEST_THREE:
+                    if (Creature* bear = ObjectAccessor::GetCreature(*me, _bearGUID))
+                    {
+                        bear->DespawnOrUnsummon();
+                        _scriptRunning = false;
+                    }
+                    break;
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:604]: indent

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 3]
```diff
@@ -422,10 +422,223 @@ class npc_prospector_remtravel : public CreatureScript
...
+                }
+            }
+
+            if (!UpdateVictim())
+                return;
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:609]: brackets

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 4]
```diff
@@ -422,10 +422,223 @@ class npc_prospector_remtravel : public CreatureScript
...
+                switch (_events.ExecuteEvent())
+                {
+                case EVENT_CHECK_FOLLOWING:
+                    Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID);
+                    if (!player || me->GetDistance2d(player) > 100.0f || me->GetMotionMaster()->GetCurrentMovementGeneratorType() != FOLLOW_MOTION_TYPE || player->GetQuestStatus(QUEST_PLAGUED_LANDS) == QUEST_STATUS_NONE)
+                    {
+                        me->DespawnOrUnsummon();
+                    }
+                    _events.ScheduleEvent(EVENT_CHECK_FOLLOWING, 1000);
+                    break;
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:500]: brackets

**Review by Nefertumm — APPROVED** [review]

File: src/server/scripts/Kalimdor/zone_darkshore.cpp (hunk 1 above)
Comment (malcrom) [reply src/server/scripts/Kalimdor/zone_darkshore.cpp:505]: No

Comment (Nefertumm) [reply src/server/scripts/Kalimdor/zone_darkshore.cpp:505]: Fine for me then, just fix indent :D

**Comment by Voxstrasza** [comment]
Tested - looks good from my PoV.  [Video](https://youtu.be/pggkyM7Dnjc)

Based on the issue, and since I accidentally lost the bear in the video, I went back and searched for a duplicate/incorrectly spawned bear and it wasn't there.  Marking as TBM

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 5]
```diff
@@ -422,10 +422,223 @@ class npc_prospector_remtravel : public CreatureScript
...
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (me->GetFaction() == FACTION_FRIENDLY)
+            {
+                _events.Update(diff);
+                switch (_events.ExecuteEvent())
+                {
+                case EVENT_CHECK_FOLLOWING:
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:493]: Please, fix this indentation

File: src/server/scripts/Kalimdor/zone_darkshore.cpp [hunk 6]
```diff
@@ -422,10 +422,223 @@ class npc_prospector_remtravel : public CreatureScript
...
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (_scriptRunning)
+            {
+                _events.Update(diff);
+                switch (_events.ExecuteEvent())
+                {
+                case EVENT_POST_QUEST_ONE:
```
Comment (Nefertumm) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:583]: same here

File: src/server/scripts/Kalimdor/zone_darkshore.cpp (hunk 3 above)
Comment (Nyeriah) [inline src/server/scripts/Kalimdor/zone_darkshore.cpp:609]:
```suggestion
            if (!UpdateVictim())
            {
                return;
            }
```

**Pull request merged**